parsing/location.mli

```diff
@@ -92,71 +92,125 @@ val show_filename: string -> string
 val print_filename: formatter -> string -> unit
 
 val print_loc: formatter -> t -> unit
-
-val default_printer : formatter -> t -> unit
-val printer : (formatter -> t -> unit) ref
-val print: formatter -> t -> unit
-val print_compact: formatter -> t -> unit
+val print_locs: formatter -> t list -> unit
 
 
 (** {1 Toplevel-specific location highlighting} *)
 
-val highlight_locations: formatter -> t list -> bool
+val highlight_terminfo:
+  Lexing.lexbuf -> formatter -> t list -> unit
+
+val highlight_dumb:
+  print_chars:bool -> Lexing.lexbuf -> formatter -> t list -> unit
+
+val highlight_locations:
+  formatter -> t list -> unit
+
+
+(** {1 Reporting errors and warnings} *)
+
+type msg = (Format.formatter -> unit) loc
+
+val msg: ?loc:t -> ('a, Format.formatter, unit, msg) format4 -> 'a
+
+type report_kind =
+  | Report_error
+  | Report_warning of int
+  | Report_warning_as_error of int
+
+type report = {
+  kind : report_kind;
+  main : msg;
+  sub : msg list;
+}
+
+type report_printer = {
+  (* The entry point *)
+  pp : report_printer ->
+    Format.formatter -> report -> unit;
+
+  pp_report_kind : report_printer -> report ->
+    Format.formatter -> report_kind -> unit;
+  pp_main_loc : report_printer -> report ->
+    Format.formatter -> t -> unit;
+  pp_main_txt : report_printer -> report ->
+    Format.formatter -> (Format.formatter -> unit) -> unit;
+  pp_submsgs : report_printer -> report ->
+    Format.formatter -> msg list -> unit;
+  pp_submsg : report_printer -> report ->
+    Format.formatter -> msg -> unit;
+  pp_submsg_loc : report_printer -> report ->
+    Format.formatter -> t -> unit;
+  pp_submsg_txt : report_printer -> report ->
+    Format.formatter -> (Format.formatter -> unit) -> unit;
+}
+(** A printer for [report]s, defined using open-recursion.
+    The goal is to make it easy to define new printers by re-using code from
+    existing ones.
+*)
+
+val batch_mode_printer: report_printer
+val toplevel_printer:
+  highlight:(formatter -> t list -> unit) ->
+  report_printer
+
+val print_report: formatter -> report -> unit
+(** Display an error or warning report. *)
 
-val show_code_at_location: formatter -> Lexing.lexbuf -> t list -> unit
+val report_printer: (unit -> report_printer) ref
+(** Hook for redefining the printer of reports. *)
+
+val default_report_printer: unit -> report_printer
+(** Original report printer for use in hooks. *)
```

Comment:
I understand why it is that way, but still find it a bit unpleasant that `default_report_printer` is `unit -> report_printer` instead of `report_printer` (and likewise for the `ref` above).

I don't have any nice alternative though (that is: I don't think saying that `batch_mode_printer` is the default and that the toplevel should update `report_printer` would be a good decision).

Reply:
Good point, I agree it is a bit unpleasant.

What do you think of this alternative proposition: 
`default_report_printer` is a `report_printer` (and likewise for the `ref`). Then, `default_report_printer` is defined as a new `report_printer` named `compiler_printer`, which does the autodetection and dispatches between `batch_mode_printer` and `toplevel_printer`.

I think it makes sense, and possibly nicer than the current proposition.

Reply:
Do you mean: each of the functions inside the record would do the dispatch?
That seems fine yes.

Reply:
Hum actually I spoke too fast, I was thinking of doing the dispatch at the entry point, but this does not make sense since you have to also export the other functions. Dispatching at each function would kinda work I guess but it is not very nice.

Reply:
Indeed. Do as you see fit (not doing anything is also an acceptable choice)

Reply:
I think I will just add a comment to explain the purpose of the `unit -> ..`.